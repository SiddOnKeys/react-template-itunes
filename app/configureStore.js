/**
 * Create the store with dynamic reducers
 */

import { createStore, applyMiddleware, compose } from 'redux';
import createSagaMiddleware from 'redux-saga';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import createReducer from './create-root-reducer';
import { createInjectorsEnhancer } from 'redux-injectors';
import rootSaga from './root-saga';

// redux persist configuration
const persistConfig = {
  version: 1,
  key: 'root',
  storage,
  blacklist: ['router'] // Don't persist router state
};

/**
 * Configure Redux DevTools enhancer
 * @returns {Function} Composed enhancer function
 */
const configureDevTools = () => {
  if (
    process.env.NODE_ENV !== 'production' &&
    typeof window === 'object' &&
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
  ) {
    return window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({
      trace: true,
      traceLimit: 25
    });
  }
  return compose;
};

/**
 * Configure Saga middleware and monitor
 * @returns {Object} Saga middleware instance
 */
const configureSaga = () => {
  const sagaOptions = {};

  if (process.env.NODE_ENV !== 'production' && typeof window === 'object' && window.__SAGA_MONITOR_EXTENSION__) {
    Object.assign(sagaOptions, { sagaMonitor: window.__SAGA_MONITOR_EXTENSION__ });
  }

  return createSagaMiddleware(sagaOptions);
};

/**
 * Configure store enhancers
 * @param {Function} sagaMiddleware - The saga middleware
 * @param {Function} runSaga - The saga runner function
 * @returns {Array} Array of enhancers
 */
const configureEnhancers = (sagaMiddleware, runSaga) => {
  const middlewares = [sagaMiddleware];
  const enhancers = [applyMiddleware(...middlewares)];

  const injectEnhancer = createInjectorsEnhancer({
    createReducer,
    runSaga
  });

  return [...enhancers, injectEnhancer];
};

/**
 * Create store extensions
 * @param {Object} store - Redux store
 * @param {Function} runSaga - Saga middleware run function
 * @returns {Object} Store with extensions
 */
const createStoreExtensions = (store, runSaga) => ({
  ...store,
  runSaga,
  injectedReducers: {},
  injectedSagas: {}
});

/**
 * Configures and creates the Redux store with middleware, enhancers, and support for hot reloading.
 * It also sets up Redux Saga and the Redux DevTools extension.
 *
 * @param {Object} [initialState={}] - The initial state of the store.
 * @returns {{ store: Object, persistor: Object }} An object containing the Redux store and the persistor for redux-persist.
 */
export default function configureStore(initialState = {}) {
  const persistedReducer = persistReducer(persistConfig, createReducer());
  const sagaMiddleware = configureSaga();
  const composeEnhancers = configureDevTools();
  const enhancers = configureEnhancers(sagaMiddleware, sagaMiddleware.run);

  const store = createStore(persistedReducer, initialState, composeEnhancers(...enhancers));

  const storeWithExtensions = createStoreExtensions(store, sagaMiddleware.run);
  const persistor = persistStore(storeWithExtensions);

  // Run the root saga
  sagaMiddleware.run(rootSaga);

  // Make reducers hot reloadable
  if (module.hot) {
    module.hot.accept('./create-root-reducer', () => {
      storeWithExtensions.replaceReducer(createReducer(storeWithExtensions.injectedReducers));
    });
  }

  return { store: storeWithExtensions, persistor };
}
