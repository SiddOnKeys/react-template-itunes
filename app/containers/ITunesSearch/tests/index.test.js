import React from 'react';
import { render, screen, fireEvent, act } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { ITunesSearch, mapDispatchToProps } from '../index';
import { searchTracks, clearTracks } from '../actions';

// Mock the styles import to avoid CSS module issues in tests
jest.mock('../styles/Container.css', () => ({
  mainContainer: 'mainContainer',
  searchContainer: 'searchContainer',
  title: 'title',
  errorMessage: 'errorMessage',
  resultsContainer: 'resultsContainer'
}));

jest.mock('../styles/Grid.css', () => ({
  gridContainer: 'gridContainer',
  gridItem: 'gridItem',
  emptyMessage: 'emptyMessage'
}));

jest.mock('@app/components/SongCard/styles.css', () => ({
  card: 'card',
  cardMedia: 'cardMedia',
  cardContent: 'cardContent',
  contentWrapper: 'contentWrapper',
  trackTitle: 'trackTitle',
  artistText: 'artistText',
  genreText: 'genreText',
  buttonsContainer: 'buttonsContainer',
  actionButton: 'actionButton',
  dialog: 'dialog',
  dialogTitle: 'dialogTitle',
  dialogContent: 'dialogContent',
  dialogLabel: 'dialogLabel',
  dialogValue: 'dialogValue',
  dialogCloseButton: 'dialogCloseButton'
}));

// Mock react-router-dom
jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: jest.fn()
  })
}));

// Mock debounce to execute immediately in tests
jest.mock('lodash/debounce', () => (fn) => {
  const debounced = (...args) => fn(...args);
  debounced.cancel = jest.fn();
  return debounced;
});

// Mock Audio API
const mockAudio = {
  play: jest.fn(),
  pause: jest.fn(),
  addEventListener: jest.fn(),
  removeEventListener: jest.fn()
};

global.Audio = jest.fn(() => mockAudio);

describe('<ITunesSearch />', () => {
  let store;
  const mockStore = configureStore([]);

  beforeEach(() => {
    store = mockStore({
      itunesSearch: {
        tracks: [],
        loading: false,
        error: null,
        query: ''
      }
    });
    jest.clearAllMocks();
  });

  const renderComponent = (props = {}) => {
    const defaultProps = {
      dispatchSearchTracks: jest.fn(),
      dispatchClearTracks: jest.fn(),
      tracks: [],
      loading: false,
      error: null,
      savedQuery: '',
      ...props
    };

    return render(
      <Provider store={store}>
        <ITunesSearch {...defaultProps} />
      </Provider>
    );
  };

  it('should render and match the snapshot', () => {
    const { container } = renderComponent();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should clear tracks on mount if no saved query exists', () => {
    renderComponent();
    expect(defaultProps.dispatchClearTracks).toHaveBeenCalled();
  });

  it('should not clear tracks on mount if saved query exists', () => {
    renderComponent({ savedQuery: 'test query' });
    expect(defaultProps.dispatchClearTracks).not.toHaveBeenCalled();
  });

  it('should update input value and trigger search on change', () => {
    renderComponent();
    const input = screen.getByRole('textbox');

    fireEvent.change(input, { target: { value: 'test search' } });

    expect(input.value).toBe('test search');
    expect(defaultProps.dispatchSearchTracks).toHaveBeenCalledWith('test search');
  });

  it('should clear search when input is cleared', () => {
    renderComponent({ tracks: mockTracks });
    const input = screen.getByRole('textbox');
    const clearButton = screen.getByTestId('clear-button');

    fireEvent.click(clearButton);

    expect(input.value).toBe('');
    expect(defaultProps.dispatchClearTracks).toHaveBeenCalled();
  });

  it('should display error message when error exists', () => {
    const error = { message: 'Test error' };
    renderComponent({ error });

    expect(screen.getByTestId('error-message')).toHaveTextContent('Error: Test error');
  });

  it('should trigger immediate search when search button is clicked', () => {
    renderComponent();
    const input = screen.getByRole('textbox');
    const searchButton = screen.getByTestId('search-button');

    fireEvent.change(input, { target: { value: 'test search' } });
    fireEvent.click(searchButton);

    expect(defaultProps.dispatchSearchTracks).toHaveBeenCalledWith('test search');
  });

  it('should show loading state', () => {
    renderComponent({ loading: true });
    expect(screen.getByRole('progressbar')).toBeInTheDocument();
  });

  it('should show empty message when no tracks', () => {
    renderComponent();
    expect(screen.getByText('Search for tracks to see results')).toBeInTheDocument();
  });

  it('should render tracks in grid', () => {
    const tracks = [
      {
        trackId: 1,
        trackName: 'Test Track',
        artistName: 'Test Artist',
        artworkUrl100: 'test.jpg',
        previewUrl: 'test.mp3',
        primaryGenreName: 'Pop',
        collectionName: 'Test Album',
        releaseDate: '2024-01-01'
      }
    ];

    renderComponent({ tracks });
    expect(screen.getByText('Test Track')).toBeInTheDocument();
    expect(screen.getByText('Test Artist')).toBeInTheDocument();
    expect(screen.getByText('Pop')).toBeInTheDocument();
  });

  describe('mapDispatchToProps', () => {
    it('should map searchTracks action', () => {
      const dispatch = jest.fn();
      const props = mapDispatchToProps(dispatch);
      const query = 'test';

      props.dispatchSearchTracks(query);
      expect(dispatch).toHaveBeenCalledWith(searchTracks(query));
    });

    it('should map clearTracks action', () => {
      const dispatch = jest.fn();
      const props = mapDispatchToProps(dispatch);

      props.dispatchClearTracks();
      expect(dispatch).toHaveBeenCalledWith(clearTracks());
    });
  });
});
