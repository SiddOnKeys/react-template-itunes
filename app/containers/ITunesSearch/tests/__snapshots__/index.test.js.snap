// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ITunesSearch /> should render and match the snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 css-ee20y-MuiPaper-root"
  >
    <h1>
      iTunes Search
    </h1>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-adornedEnd css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
      data-testid="search-input"
    >
      <input
        class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
        data-testid="search-input-field"
        placeholder="Search for tracks..."
        type="text"
        value=""
      />
      <div
        class="MuiInputAdornment-root MuiInputAdornment-positionEnd css-1laqsz7-MuiInputAdornment-root"
      >
        <button
          aria-label="Search tracks"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          data-testid="search-button"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="SearchIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-ihdtdm"
        >
          <span
            class="notranslate"
          >
            â€‹
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 css-1dn6hbf-MuiPaper-root"
  >
    <div
      class="gridContainer"
    >
      <div
        class="emptyMessage"
        data-testid="no-results"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          Search for tracks to see results
        </p>
      </div>
    </div>
  </div>
</div>
`;
